src/constants/types.rs

```diff
@@ -513,6 +514,46 @@ impl ResourceType {
 
 js_deserializable!(ResourceType);
 
+/// Translates market resource types which can include both `RESOURCE_*`
+/// and `INTERSHARD_RESOURCES` constants.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum MarketResourceType {
+    Resource(ResourceType),
+    IntershardResource(IntershardResourceType),
+}
+
+impl<'de> Deserialize<'de> for MarketResourceType {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: Cow<'de, str> = Cow::deserialize(d)?;
+
+        ResourceType::from_str(&s)
+            .map(|ty| MarketResourceType::Resource(ty))
+            .or(IntershardResourceType::from_str(&s)
+                .map(|ty| MarketResourceType::IntershardResource(ty)))
+            .map_err(|_| {
+                D::Error::invalid_value(
+                    Unexpected::Str(&s),
+                    &"a known constant string in RESOURCES_ALL or INTERSHARD_RESOURCES",
+                )
+            })
+    }
+}
+
+impl Serialize for MarketResourceType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            MarketResourceType::Resource(ty) => ty.serialize(serializer),
+            MarketResourceType::IntershardResource(ty) => ty.serialize(serializer),
```

Comment:
Are you sure this `Serialize` impl actually matches the `Deserialize` one?

As I recall, `ResourceType` and `IntershardResourceType` both implement `Serialize` / `Deserialize` by serializing to our fake numbers, not to string representations. Since the `Deserialize` impl for `MarketResourceType` depends on string representations, I think we probably want to serialize to strings as well here. Either that, or expand the `Deserialize` impl to accept the integers too?

Reply:
Good catch - needed to be updated to have both the magic-int parse on the normal deserialize, as well as adding the `MarketResourceType::deserialize_from_str` function for processing API value returns.  I made it yet-another-exhaustive-match-of-the-whole-list-of-resources which is a little painful, but I couldn't figure out a more clever way to try deserializing as one type then the other; let me know if it should be improved!

Reply:
Looks good to me! There might be something we could do with trying to deserialize ResourceType from the  `u16` first, then if that fails trying MarketResourceType, but I'm not entirely sure if that would be easier/better or how possible it'd be.

Sorry about not getting back to this immediately!